ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/GSM.hs

```diff
@@ -0,0 +1,698 @@
+
+toJudgment :: ModelState -> LedgerStateJudgement
+toJudgment = \case
+    ModelTooOld        -> TooOld
+    ModelYoungEnough{} -> YoungEnough
+
+toMarker :: LedgerStateJudgement -> MarkerState
+toMarker = \case
+    TooOld      -> Absent
+    YoungEnough -> Present
+
+-- | Update the 'mState', assuming that's the only stale field in the given
+-- 'Model'
+fixupModelState :: Model r -> Model r
+fixupModelState model =
+    case st of
+        ModelTooOld | caughtUp ->
+            -- ASSUMPTION This new state was /NOT/ incurred by the 'TimePasses'
+            -- command.
+            --
+            -- Therefore the current clock is necessarily the correct timestamp
+            -- to record.
+            model { mState = ModelYoungEnough clk }
+        ModelYoungEnough timestamp | fellBehind timestamp ->
+            -- ASSUMPTION The 'TimePasses' command incurred this new state.
+            --
+            -- It's possible for the node to instantly return to CaughtUp, but
+            -- that might have happened /during/ the 'TimePasses' command, not
+            -- only when it ends.
+            --
+            -- Therefore the age limit of the selection is the correct
+            -- timestamp to record, instead of the current clock (ie when the
+            -- 'TimePasses' ended).
+            --
+            -- NOTE Superficially, in the real implementation, the Diffusion
+            -- Layer should be discarding all peers when transitioning from
+            -- CaughtUp to OnlyBootstrap. However, it would be plausible for an
+            -- implementation to retain any bootstrap peers it happened to
+            -- have, so the idiosyncratic behavior of the system under test in
+            -- this module is not totally irrelevant.
+            if caughtUp
+            then model { mState = ModelYoungEnough (timestamp' timestamp) }
+            else model { mState = ModelTooOld }
+        _ ->
+            model
+  where
+    Model {
+        mCandidates = cands
+      ,
+        mClock = clk
+      ,
+        mIdlers = idlers
+      ,
+        mSelection = sel
+      ,
+        mState = st
+      } = model
+
+    caughtUp             = some && allIdling && all ok cands
+    fellBehind timestamp = tooOld && notThrashing timestamp
+
+    some = 0 < Map.size cands
+
+    allIdling = idlers == Map.keysSet cands
+
+    ok cand =
+        GSM.WhetherCandidateIsBetter False == candidateOverSelection sel cand
+
+    fortiethBirthday = SI.addTime ageLimit (onset sel)
+    tooOld           = fortiethBirthday < clk   -- NB 'boringDur' prevents ==
+
+    notThrashing timestamp =
+        SI.addTime thrashLimit timestamp < clk   -- NB 'boringDur' prevents ==
+
+    -- The /last/ time the node instantaneously visited OnlyBootstrap during
+    -- the 'TimePasses' command.
+    timestamp' timestamp =
+        foldl max fortiethBirthday
+      $ filter (< clk)   -- NB 'boringDur' prevents ==
+      $ iterate (SI.addTime thrashLimit) timestamp
```

Comment:
I *think* I understand what the purpose of this is, but I was surprised that replacing this definition by
```haskell
    timestamp' timestamp = fortiethBirthday
```
also results in the tests passing just fine.

Reply:
Thank you very much for poking around these parts. Along with your

> However, the tests often just hang indefinitely (for example, for --quickcheck-replay=249156). Is there a simple explanation for this?

comment, I ended up finding a nest of problems around this definition! (The planned task of labeling QC seeds would also have revealed this, FTR.)

- I hard coded a 10^-9 for pico, instead of 10^-12 🤦 
- This `filter` should be a `takeWhile`.

The second one explains the hang. The first one explains why `boringDur` had rendered this dead-code.

There were more problems. I _think_ I've resolved them (at the cost of further complexity in the `precondition` and `generator` functions, sadly), but I'm still seeing failures. After a few hours, I think those are due to an `io-sim` bug 😢. I've pinged on the Slack and will escalate to an Issue once someone confirms (or else explains my misunderstanding in such a way that I can fix my code).